File: model/src/main/scala/ru/mipt/acsl/decode/model/types/DecodeType.scala
package ru.mipt.acsl.decode.model.types

import ru.mipt.acsl.decode.model.naming.{ElementName, Fqn, Namespace}
import ru.mipt.acsl.decode.model.{LocalizedString, Referenceable}

/**
  * @author Artem Shein
  */
trait DecodeType extends Referenceable {

  def namespace: Namespace

  def namespace_=(ns: Namespace): Unit

  def info: LocalizedString = alias.map(_.info).getOrElse(LocalizedString.empty)

  def alias: Option[Alias.Type]

  def systemName: String

  def typeParameters: Seq[ElementName]

  def isGeneric: Boolean = typeParameters.nonEmpty

  def nameOrSystemName: String = alias.map(_.name.asMangledString).getOrElse(systemName)

  override def toString: String = s"{alias: ${alias.toString}}"

  def fqn: Option[Fqn] = alias.map(a => Fqn(namespace.fqn.parts :+ a.name))

  def isUnit: Boolean = sys.error("not implemented")

  def isArray: Boolean = sys.error("not implemented")

  def isNative: Boolean = this match {
    case _: NativeType => true
    case _ => false
  }

  def isOrType: Boolean = fqn.contains(Fqn.Or)

  def isOptionType: Boolean = fqn.contains(Fqn.Option)

  def isVaruintType: Boolean = fqn.contains(Fqn.Varuint)

  def nameOption: Option[ElementName] = alias.map(_.name)

}

File: model/src/main/java/ru/mipt/acsl/decode/model/types/DecodeType.java
package ru.mipt.acsl.decode.model.types;

import ru.mipt.acsl.decode.model.HasNamespace;
import ru.mipt.acsl.decode.model.MayHaveAlias;
import ru.mipt.acsl.decode.model.Referenceable;
import ru.mipt.acsl.decode.model.ReferenceableVisitor;
import ru.mipt.acsl.decode.model.naming.ElementName;
import ru.mipt.acsl.decode.model.naming.Fqn;
import ru.mipt.acsl.decode.model.registry.Language;

import java.util.*;

/**
 * Created by metadeus on 06.06.16.
 */
public interface DecodeType extends Referenceable, HasNamespace, MayHaveAlias {

    default Map<Language, String> info() {
        return alias()
                .map(Alias::info)
                .orElseGet(HashMap::new);
    }

    @Override
    Optional<Alias> alias();

    String systemName();

    List<ElementName> typeParameters();

    default boolean isGeneric() {
        return !typeParameters().isEmpty();
    }

    default String nameOrSystemName() {
        return alias()
                .map(a -> a.name().mangledNameString())
                .orElse(systemName());
    }

    default Optional<Fqn> fqn() {
        return alias()
                .map(a -> {
                    List<ElementName> parts = new ArrayList<>(namespace().fqn().getParts());
                    parts.add(a.name());
                    return Fqn.newInstance(parts);
                });
    }

    default boolean isUnit() {
        throw new RuntimeException("not implemented");
    }

    default boolean isArray() {
        throw new RuntimeException("not implemented");
    }

    default boolean isNative() {
        return this instanceof NativeType;
    }

    default boolean isOrType() {
        return fqn().map(f -> f.equals(Fqn.OR)).orElse(false);
    }

    default boolean isOptionType() {
        return fqn().map(f -> f.equals(Fqn.OPTION)).orElse(false);
    }

    default boolean isVaruintType() {
        return fqn().map(f -> f.equals(Fqn.VARUINT)).orElse(false);
    }

    default Optional<ElementName> nameOption() {
        return alias().map(Alias::name);
    }

    @Override
    default <T> T accept(ReferenceableVisitor<T> visitor) {
        return visitor.visit(this);
    }

}
